### PR TITLE
2452 name merging name link and invoice

### LINKED ISSUE #2452: Name merging: name_link and invoice
## Is your feature request related to a problem? Please describe 👀

Open mSupply needs to support name merging using a "link" table. 

## Describe the solution you'd like 🎁
Base: [feature/merging](https://github.com/msupply-foundation/open-msupply/tree/feature/merging)

Using changes in the [item_link + stock_line PR](https://github.com/msupply-foundation/open-msupply/pull/2547/files#diff-63618ddd9539c21a7c93d1c9432eb912541ea89aa94f0b73c7052550ba01770f) as a guide:
- [x] Create a new table `name_link`; `name_link.id` String; `name_link.name_id` String FK to `name.id`
- [x] Migration populating `name_link.id == name_link.name_id == name.id` for each `name` record
- [ ] Migration adding `invoice.name_link_id` and removing `invoice.name_id`
- [ ] Migration fixing up any sql views
- [ ] Update `invoice` at the repository layer to return related `name` records based on `invoice.name_link_id` rather than `invoice.name_id`
- [ ] Fix all the tests 🧪 

### Moneyworks Jobcode 🧰

OMS:MERGE

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #2452

# 👩🏻‍💻 What does this PR do? 
- Migration to add `name_link` table. 
- Migration to remove `invoice.name_id` and add `invoice.name_link_id`
- Migrations to recreate (for sqlite since can't drop views) and update (for postgres) triggers for invoice and invoice lines. 
- Fixed tests that rely on previous `invoice.name_id` 

# 🧪 How has/should this change been tested? 
- [ ] Run tests.
- [ ] Make sure invoice functionality is still working as intended.

### CODE REVIEW

File: server/repository/src/db_diesel/name_link_row.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    pub async fn insert_one(&self, row: &NameLinkRow) -> Result<(), RepositoryError> {
+        diesel::insert_into(name_link)
+            .values(row)
+            .execute(&self.connection.connection)?;
+        Ok(())
+    }
+
+    pub async fn find_one_by_id(&self, item_link_id: &str) -> Result<NameLinkRow, RepositoryError> {
```
Comment (andreievg) [inline server/repository/src/db_diesel/name_link_row.rs:57]: name_link_id

File: server/repository/src/db_diesel/invoice_row.rs [hunk 2]
```diff
@@ -222,7 +224,7 @@ impl<'a> OutboundShipmentRowRepository<'a> {
             .filter(
                 type_
                     .eq(InvoiceRowType::OutboundShipment)
-                    .and(name_id.eq(name)),
+                    .and(name_link_id.eq(name)),
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/invoice_row.rs:227]: This seems suss to me - should really be on `name_link::name_id` in some manner and this doesn't seem like that, will have to look again in rust analyser...

All that said, the only place this function is called is in a test anyway, but I can see need to see all outbound shipments for a particular customer/patient. Though... I am curious why there is no equivalent `InboundShipmentRowRepository`? Is `OutboundShipmentRowRepository` actually not used outside of tests? Can we just remove it? I have a suspicion all shipments are just queried through the base invoice repository really.

**Review by Chris-Petty — CHANGES_REQUESTED** [review]
Pretty much good to go! Just found one thing seeming off...

File: server/repository/src/migrations/v1_06_00/invoice_add_name_link_id.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        ALTER TABLE invoice ADD CONSTRAINT invoice_name_link_id_fkey FOREIGN KEY (name_link_id) REFERENCES name_link(id);
+        "#,
+    )?;
+    sql!(
+        connection,
+        r#"
+        CREATE OR REPLACE FUNCTION upsert_invoice_changelog()
+        RETURNS trigger AS
+        $$ BEGIN
+            INSERT INTO changelog (table_name, record_id, row_action, name_id, store_id)
```
Comment (Chris-Petty) [inline server/repository/src/migrations/v1_06_00/invoice_add_name_link_id.rs:24]: Noting that we're setting `name_link_id` onto `name_id`. Chatted to Andrei, to keep the trigger overhead low it's good to not doing a join on `name_link` to get the `name_id`, better to change `ChangelogRepository` to do the join. Good for another issue, I'll make it.

File: server/repository/src/migrations/v1_06_00/invoice_add_name_link_id.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
+use crate::{migrations::sql, StorageConnection};
+
+#[cfg(feature = "postgres")]
+pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/migrations/v1_06_00/invoice_add_name_link_id.rs:4]: Whatever you did seems to have fixed it hahahah

**Review by Chris-Petty — APPROVED** [review]
LGTM! ✨ I fixed conflicts with base, all tests passin'

**Pull request merged**